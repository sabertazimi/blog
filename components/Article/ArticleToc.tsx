'use client'

import { useEffect, useState } from 'react'
import styles from './ArticleToc.module.css'
import Anchor from '@/components/Anchor'

interface Props {
  slug: string
  offset?: number
}

interface TocItem {
  key: string | number
  href: string
  title: string
}

function ArticleToc({ slug, offset = 150 }: Props): JSX.Element {
  const [tocItems, setTocItems] = useState<TocItem[]>([])

  useEffect(() => {
    const items = document.querySelectorAll('h2.ant-typography')
    const tocItems = Array.from(items).map(item => ({
      key: `#${item.id}`,
      href: `#${item.id}`,
      title: item.textContent,
    })) as TocItem[]
    setTocItems(tocItems)
  }, [slug])

  return (
    <div className={styles.toc}>
      <Anchor offsetTop={offset} items={tocItems} />
    </div>
  )
}

export default ArticleToc
