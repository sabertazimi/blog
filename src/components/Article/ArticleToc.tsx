import { CloseOutlined, MenuFoldOutlined } from '@ant-design/icons';
import { Button, Drawer } from 'antd';
import classNames from 'classnames';
import React, { useCallback, useState } from 'react';
import * as styles from './ArticleToc.module.css';

interface Props {
  toc: string;
}

const ArticleToc = ({ toc }: Props): JSX.Element => {
  const [tocVisible, setTocVisible] = useState(false);

  const handleClick = useCallback(
    () => setTocVisible(tocVisible => !tocVisible),
    []
  );

  return (
    <div
      className={classNames(
        'fixed hidden m-0 bg-transparent',
        'top-8 right-28 z-max',
        'md:block'
      )}
      role="navigation"
    >
      <Button
        className="button-primary"
        shape="circle"
        icon={tocVisible ? <CloseOutlined /> : <MenuFoldOutlined />}
        size="large"
        onClick={handleClick}
      />
      <Drawer
        className="font-extrabold text-dark"
        title="Table of Contents"
        placement="right"
        closable={false}
        onClose={handleClick}
        visible={tocVisible}
      >
        <div
          className={classNames('transition transform-gpu', styles.toc, {
            'scale-100': tocVisible,
            'scale-0': !tocVisible,
          })}
          dangerouslySetInnerHTML={{ __html: toc }}
          onClick={handleClick}
        />
      </Drawer>
    </div>
  );
};

export default ArticleToc;
