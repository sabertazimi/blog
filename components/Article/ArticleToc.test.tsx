import { render } from '@utils'
import ArticleToc from './ArticleToc'

describe('ArticleToc', () => {
  test('should render correctly (snapshot)', () => {
    const { container } = render(<ArticleToc slug="slug" />)

    expect(container).toMatchSnapshot()
  })
})
