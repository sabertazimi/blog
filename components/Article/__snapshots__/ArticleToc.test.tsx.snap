// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleToc should render correctly (snapshot) 1`] = `
<div>
  <div
    class="toc"
  >
    <div>
      <div
        class=""
      >
        <div
          class="css-dev-only-do-not-override-j0nf2s ant-anchor-wrapper"
          style="max-height: calc(100vh - 150px);"
        >
          <div
            class="ant-anchor"
          >
            <span
              class="ant-anchor-ink"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
